### DISCUSSION TITLE
Upgrade to Use AHK v2

### DISCUSSION THREAD

**Discussion by chrstophr-wltrs** [Lailloken/Exile-UI#235] [Ideas & Feedback]
I'm generally an advocate for using the most secure and feature-rich version of a technology whenever and wherever possible. LailloUI depends on AHK v1.1 which is now deprecated, and I think there will be major performance and DevX improvements by using something more contemporary. 

From what I understand, the major difference is that v1 uses lots of labels/subroutines, whereas v2 prefers functions. It's mostly a syntax difference, honestly.

I'd originally planned on opening up a pull request and doing it myself, but based on Lailloken's README, I don't think that would be his preference.

> I will not be taking direct code input or optimizations from people because learning-by-doing and problem-solving is what makes this project fun for me.

Additionally, this note was interesting to me:

> having to go through, double-check, and test other people's code sounds like a chore to me.

This is almost exactly why unit tests exist: so that you don't need to double-check code. I'm not sure it's necessary for a project this size, but I'd definitely recommend looking into automated testing in the future. 1-2 hours of test writing magnifies the speed at which a project can progress near exponentially.

**Comment by Lailloken** [comment]
Thanks for your input. I haven't transitioned the project to v2 yet for the simple reason that I didn't realize it was out of beta until someone mentioned it earlier this week. I missed it because it had been in beta for years and I don't really do any research or learning when I'm not actively working on my project.

My "flying solo" approach laid out in the readme is based on the fact that I don't really consider myself a developer. I treat this project more like another game in my library of games to play in my spare time. If I have some free time and nobody is available for a multi-player session and I also don't feel like playing anything single-player, I turn to this project to keep myself busy/entertained.

The "chore" part is more of a cop-out statement. To continue the game analogy, having (more experienced) people working on this project with me could potentially be a threat to my enjoyment -- the sole reason I started the project. Imagine playing a brand new, GOTY coop-game with a friend, but they currently have more spare time than you. Finding out they're secretly playing a separate solo-campaign can be devastating to your enjoyment of the game if it manifests as them rushing through your coop-playthrough, blasting every enemy before they really appear on screen, knowing every secret, and making it their own ride where you are merely a passive passenger.

It would be cool if this project had the best features around, or if it revolutionized PoE gameplay, but that is not my intent. As mentioned in the readme, I don't see it as a product -- the journey is what matters to me, not the destination. I've already encountered people who find this stance strange and I always took it as elitism. But this has become more frequent, and I don't know what to think anymore.

**Reply by chrstophr-wltrs** [reply]
It's no problem at all, thanks for sharing some of your perspective and reasoning with me.

> I didn't realize it was out of beta

I totally get that. Time seriously feels like it's flying by the last few years. I've been re-installing AHK as I get new machines, so I didn't even realize that I'd transitioned to using a newer version. 

>  Flying solo, coding as a game, and "getting carried"

I completely sympathize, at least in regards to gaming. I really hate the sensation of "getting carried" in any game I play; League of Legends and Path of Exile come to mind. 

> I always took it as elitism.

I've definitely encountered developers with ego issues, which can make them difficult to work with. It's part of the reason why I phrased my initial request the way that I did-- I have coworkers who are highly resistant to the idea of upgrading technologies for existing projects simply because they were irritated by having to learn new things. 

I apologize if my original post came across as condescending or overly demanding, that was definitely not my intention. I am seriously impressed by the functionality and reliability of LailloUI, and it's clear I'm not the only one. A recent [video](https://www.youtube.com/watch?v=E61Mbuwr6Og) by a notable PoE content creator highlighting your project has nearly 100k views at the time of writing. 

> I've already encountered people who find this stance strange [ ... ] I don't know what to think anymore.

I guess I can only say this: most people view software development as a pathway to creating products which solve problems. Yes, the journey itself is enjoyable and fulfilling for many, but they also find satisfaction in solving the problem. If they want to keep coding, there are *always* more problems to solve. 

Whether you like it or not, the quality of the work that you've done is going to attract more users who treat the end utility as a product for solving their problem. The PoE playerbase has a higher percentage of developers than most games (see Path of Building), and so you're probably going to get more suggestions for ideas or pull requests, not because they feel you've done a poor job, but exactly the opposite-- they admire the work you've done, and want to help "solve the problem" in a different or enhanced way.

I say all this not to try and argue that one perspective is better than the other-- I don't begrudge you your enjoyment in working on this fantastic project. I guess I just wanted to clarify the other approach, and potentially outline some of the merits that merging the two systems could have.

**Reply by chrstophr-wltrs** [reply]
In regards to AHK v2, I guess my reasons for requesting this were slightly selfish in their own right: I use AHK for my own utilities and helpers, and managing 2 separate versions of the runtime is a little tedious, but I'd still really love to try out the end result of your hard work.

**Reply by chrstophr-wltrs** [reply]
If you don't mind my asking, would you be at all hurt, offended, or irritated if another developer were to create a product very similar to your own. (I wouldn't judge btw, I know that would kind of annoy me, personally.)

**Reply by Lailloken** [reply]
> I apologize if my original post came across as condescending or overly demanding

It didn't. Maybe I shouldn't have written my reply first thing in the morning, straight after getting out of bed. I just wanted to provide the full context -- half-context seemed to have been worse than no context when I explained (or tried to) my stance to some people in the past.

> would you be at all hurt, offended, or irritated if another developer were to create a product very similar to your own.

I can't say I like where this question is going, but at the same time there's nothing I could do about it. I mean, it's not like everything about my project is unique -- it's mostly my take on solving some problems/inconveniences that exist in PoE and that others have approached as well. So as long as such a product isn't a straight carbon copy of my project just for arbitrary reasons, or to spite me or my stance, I'll be fine. Even then, I would pride myself on being the obvious inspiration/source and go on with my life.

> and managing 2 separate versions of the runtime is a little tedious

Hmm, off the top of my head, in that situation I would get the portable/zip version of 1.1.x, put it somewhere out of sight, and then create a shortcut to its executable and define LLK-UI as a launch parameter. It shouldn't interfere with the system or v2 installation. But I'm not entirely sure.


Regarding v2: I'll definitely make the transition, and probably start looking into it around 2-3 weeks into the league (I don't see myself playing a crafting/gear-heavy league for too long). I was planning on rebuilding my project from the ground up anyway at some point -- working on it is starting to feel a bit like playing Jenga, and some aspects are in dire need of polish or a rework. It was born out of quarantine-boredom and shows clear signs of scatter-brain syndrome where a lack of planning becomes more and more apparent.

**Reply by chrstophr-wltrs** [reply]
> I can't say I like where this question is going, but at the same time there's nothing I could do about it.

I know, but I would want to ask anyway. I have my own projects to work on. I wasn't asking because I intended to do something like that, I was mostly curious, because I've seen it happen before. 

That tip about going with the portable version is probably what I'd end up doing. 

> working on it is starting to feel a bit like playing Jenga

😔 Such is the curse of all long-lived projects. 

![xkcd comic regarding dependencies and how abstraction makes developers reliant on poorly-understood technologies](https://imgs.xkcd.com/comics/dependency.png)
